gflownet/envs/tree.py

```diff
@@ -0,0 +1,1079 @@
+
+                    parent = self._set_stage(Stage.COMPLETE, parent)
+                    attributes[Attribute.TYPE] = NodeType.CLASSIFIER
+                    attributes[Attribute.FEATURE] = -1
+                    attributes[Attribute.THRESHOLD] = -1
+                    attributes[Attribute.CLASS] = output
+                    attributes[Attribute.ACTIVE] = Status.INACTIVE
+
+                    parents.append(parent)
+                    actions.append((ActionType.PICK_LEAF, k))
+            elif stage == Stage.FEATURE:
+                # Reverse self._pick_feature.
+                parent = state.clone()
+                attributes = parent[k]
+
+                parent = self._set_stage(Stage.LEAF, parent)
+                attributes[Attribute.FEATURE] = -1
+
+                parents.append(parent)
+                actions.append((ActionType.PICK_FEATURE, self.state[k][1].item()))
+            elif stage == Stage.THRESHOLD:
+                # Reverse self._pick_threshold.
+                parent = state.clone()
+                attributes = parent[k]
+
+                parent = self._set_stage(Stage.FEATURE, parent)
+                attributes[Attribute.THRESHOLD] = -1
+
+                parents.append(parent)
+                actions.append((ActionType.PICK_THRESHOLD, -1))
+            else:
+                raise ValueError(f"Unrecognized stage {stage}.")
+
+        return parents, actions
+
+    @staticmethod
+    def action2representative(action: Tuple) -> Tuple:
+        """
+        Replaces the continuous value of a PICK_THRESHOLD action by -1 so that it can
+        be contrasted with the action space and masks.
+        """
+        if action[0] == ActionType.PICK_THRESHOLD:
+            action = (ActionType.PICK_THRESHOLD, -1)
+        return action
+
+    def get_max_traj_length(self) -> int:
+        return self.n_nodes * Attribute.N
+
+    def _get_graph(self, graph: Optional[nx.DiGraph] = None, k: int = 0) -> nx.DiGraph:
+        """
+        Recursively convert self.state into a networkx directional graph.
+        """
+        if graph is None:
+            graph = nx.DiGraph()
+
+        attributes = self.state[k]
+        graph.add_node(k, x=attributes)
+
+        if attributes[Attribute.TYPE] != NodeType.CLASSIFIER:
+            k_left = Tree._get_left_child(k)
+            if not torch.any(torch.isnan(self.state[k_left])):
+                self._get_graph(graph, k=k_left)
+                graph.add_edge(k, k_left)
+
+            k_right = Tree._get_right_child(k)
+            if not torch.any(torch.isnan(self.state[k_right])):
+                self._get_graph(graph, k=k_right)
+                graph.add_edge(k, k_right)
+
+        return graph
+
+    def _to_pyg(self) -> pyg.data.Data:
+        """
+        Convert self.state into a PyG graph.
+        """
+        return from_networkx(self._get_graph())
+
+    @staticmethod
+    def _load_dataset(data_path):
+        from pathlib import Path
```

Comment:
Why not have this import at the top of the file?

Reply:
Just because it's only needed there, but it's a lightweight import so I will move it up.

Reply:
Done [here](https://github.com/alexhernandezgarcia/gflownet/pull/151/commits/896196df7007dc7ecea651b4ca2f3558efb46d55.